File: contracts/seth/test-outbound-lock.py
#!/usr/bin/env python3
"""
Outbound test (Seth -> Solana):
Create a withdraw/lock request and query lock-style fields including lockRequestKey.

Usage:
  cd contracts/seth
  python test-outbound-lock.py --amount-seth 10 --solana-recipient-base58 GRuCu61Pfyub9CgYjnqQGq9aeGAUKAUZ9pBEpLExjDqy
"""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path

from request_withdraw_to_solana_from_seth import (
    get_total_withdraw_requests,
    get_withdraw_request,
    parse_hex32,
    solana_pubkey_bytes_from_base58,
    query_contract_raw,
    _extract_hex_output,
    _decode_uint256_word,
)
from deploy_seth import SethClient, encode_call, require_seth_tx_ok


def q_word(host: str, port: int, from_hex: str, bridge_hex: str, fn_sig: str, request_id: int) -> str | None:
    call_hex = encode_call(fn_sig, ["uint256"], [request_id])
    raw = query_contract_raw(host, port, from_hex, bridge_hex, call_hex)
    out = _extract_hex_output(raw or "")
    if not out or len(out) < 64:
        return None
    return out[:64]


def q_lock_key(host: str, port: int, from_hex: str, bridge_hex: str, request_id: int) -> str | None:
    for fn_sig in ("lockRequestKey(uint256)", "withdrawRequestKey(uint256)"):
        w = q_word(host, port, from_hex, bridge_hex, fn_sig, request_id)
        if w:
            return "0x" + w
    return None


def main() -> int:
    p = argparse.ArgumentParser(description="Seth outbound lock/withdraw request test")
    p.add_argument("--host", default=os.environ.get("SETH_HOST", "35.197.170.240"))
    p.add_argument("--port", type=int, default=int(os.environ.get("SETH_PORT", "23001")))
    p.add_argument("--bridge", default=None, help="SethBridge address; default from deployment-info.json")
    p.add_argument("--user-key", default=os.environ.get("USER_PRIVATE_KEY") or os.environ.get("WITHDRAW_USER_PRIVATE_KEY"))
    p.add_argument("--amount-seth", type=int, default=1)
    p.add_argument("--min-susdc-raw", type=int, default=0)
    p.add_argument("--gas-limit-call", type=int, default=5_000_000)
    p.add_argument("--solana-recipient-hex", default=None)
    p.add_argument("--solana-recipient-base58", default=None)
    args = p.parse_args()

    if not args.user_key:
        raise SystemExit("ERROR: set USER_PRIVATE_KEY or --user-key")
    pk = args.user_key if args.user_key.startswith("0x") else ("0x" + args.user_key)

    if args.solana_recipient_base58:
        sol32 = solana_pubkey_bytes_from_base58(args.solana_recipient_base58)
    elif args.solana_recipient_hex:
        sol32 = parse_hex32(args.solana_recipient_hex)
    else:
        raise SystemExit("ERROR: pass --solana-recipient-hex or --solana-recipient-base58")

    dep = Path(__file__).resolve().parent / "deployment-info.json"
    if args.bridge:
        bridge = args.bridge.replace("0x", "").lower()
    elif dep.is_file():
        bridge = json.loads(dep.read_text(encoding="utf-8"))["SethBridge"].replace("0x", "").lower()
    else:
        raise SystemExit("ERROR: use --bridge or create deployment-info.json")

    client = SethClient(args.host, args.port)
    user_addr = client.get_address(pk.replace("0x", ""))
    before = get_total_withdraw_requests(args.host, args.port, user_addr, bridge)
    print(f"[outbound] before totalWithdrawRequests={before}")

    input_hex = encode_call(
        "requestWithdrawToSolanaFromSETH(bytes32,uint256)",
        ["bytes32", "uint256"],
        [sol32, args.min_susdc_raw],
    )
    txh = client.send_transaction_auto(
        pk,
        bridge,
        amount=args.amount_seth,
        gas_limit=args.gas_limit_call,
        gas_price=1,
        step=8,
        input_hex=input_hex,
    )
    if not txh:
        raise SystemExit("ERROR: send failed")
    ok, st = client.wait_for_receipt(txh)
    if not ok:
        raise SystemExit("ERROR: receipt wait timed out")
    require_seth_tx_ok("requestWithdrawToSolanaFromSETH", txh, st)
    print(f"[outbound] tx={txh} status={st}")

    after = get_total_withdraw_requests(args.host, args.port, user_addr, bridge)
    print(f"[outbound] after totalWithdrawRequests={after}")
    if after <= before:
        raise SystemExit("ERROR: request counter did not increase")

    req = get_withdraw_request(args.host, args.port, user_addr, bridge, after)
    key = q_lock_key(args.host, args.port, user_addr, bridge, after)
    print(f"[outbound] request_id={after}")
    print(f"[outbound] request={req}")
    print(f"[outbound] lockRequestKey={key}")
    print("[outbound] relayer should unlock on Solana and then mark processed on Seth")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: contracts/seth/test-outbound-lock.js
#!/usr/bin/env node
/**
 * Outbound test (Seth -> Solana) in JavaScript.
 *
 * Flow:
 * 1) User calls SethBridge.requestWithdrawToSolanaFromSETH(bytes32,uint256)
 * 2) Verify totalWithdrawRequests increments
 * 3) Query request fields + lock key (lockRequestKey/withdrawRequestKey)
 *
 * Usage:
 *   cd contracts/seth
 *   set USER_PRIVATE_KEY=0x...
 *   node test-outbound-lock.js --amount-seth 10 --solana-recipient-hex <64_hex>
 *
 * Optional:
 *   --bridge 0x...
 *   --min-susdc-raw 0
 *   --gas-limit-call 5000000
 *   --solana-recipient-base58 <pubkey>   (requires: npm i bs58)
 */

const fs = require('fs');
const path = require('path');
const createKeccakHash = require('keccak');
const secp256k1 = require('secp256k1');
const SethClient = require(path.join(__dirname, '../../relayer/sethClient.js'));

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function functionSelector(signature) {
  return createKeccakHash('keccak256').update(signature).digest('hex').slice(0, 8);
}

function uint256Hex(v) {
  return BigInt(v).toString(16).padStart(64, '0');
}

function parseArgs(argv) {
  const out = {
    host: process.env.SETH_HOST || '35.197.170.240',
    port: parseInt(process.env.SETH_PORT || '23001', 10),
    bridge: null,
    userKey: process.env.USER_PRIVATE_KEY || process.env.WITHDRAW_USER_PRIVATE_KEY || '',
    amountSeth: 1,
    minSusdcRaw: 0,
    gasLimitCall: 5_000_000,
    solanaRecipientHex: null,
    solanaRecipientBase58: null,
  };

  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    const n = argv[i + 1];
    if (a === '--host' && n) out.host = n, i++;
    else if (a === '--port' && n) out.port = parseInt(n, 10), i++;
    else if (a === '--bridge' && n) out.bridge = n, i++;
    else if (a === '--user-key' && n) out.userKey = n, i++;
    else if (a === '--amount-seth' && n) out.amountSeth = parseInt(n, 10), i++;
    else if (a === '--min-susdc-raw' && n) out.minSusdcRaw = parseInt(n, 10), i++;
    else if (a === '--gas-limit-call' && n) out.gasLimitCall = parseInt(n, 10), i++;
    else if (a === '--solana-recipient-hex' && n) out.solanaRecipientHex = n, i++;
    else if (a === '--solana-recipient-base58' && n) out.solanaRecipientBase58 = n, i++;
  }

  return out;
}

function normalizePrivateKey(pk) {
  if (!pk) throw new Error('Missing user private key (--user-key or USER_PRIVATE_KEY)');
  return pk.startsWith('0x') ? pk : `0x${pk}`;
}

function normalizeBridgeAddress(addr) {
  const h = String(addr || '').trim().toLowerCase().replace(/^0x/, '');
  if (!/^[0-9a-f]{40}$/.test(h)) throw new Error(`Invalid bridge address: ${addr}`);
  return h;
}

function parseRecipientHex64(v) {
  const h = String(v || '').trim().toLowerCase().replace(/^0x/, '');
  if (!/^[0-9a-f]{64}$/.test(h)) {
    throw new Error('solana recipient hex must be 64 hex chars');
  }
  return h;
}

function parseRecipientBase58(v) {
  try {
    const bs58 = require('bs58');
    const raw = bs58.decode(String(v).trim());
    if (raw.length !== 32) throw new Error(`decoded length ${raw.length} != 32`);
    return Buffer.from(raw).toString('hex');
  } catch (e) {
    throw new Error(`base58 parse failed (${e.message}); install bs58: npm i bs58`);
  }
}

function loadBridgeAddress(override) {
  if (override) return normalizeBridgeAddress(override);
  const depPath = path.join(__dirname, 'deployment-info.json');
  if (!fs.existsSync(depPath)) {
    throw new Error('Missing --bridge and deployment-info.json not found');
  }
  const j = JSON.parse(fs.readFileSync(depPath, 'utf8'));
  if (!j.SethBridge) throw new Error('deployment-info.json missing SethBridge');
  return normalizeBridgeAddress(j.SethBridge);
}

function deriveAddressFromPrivateKey(client, privateKeyHex) {
  const pk = Buffer.from(privateKeyHex.replace(/^0x/, ''), 'hex');
  const pub = secp256k1.publicKeyCreate(pk, false);
  return client.deriveAddressFromPubkey(pub).replace(/^0x/, '');
}

async function waitReceipt(client, txHash, maxAttempts = 180, intervalMs = 2000) {
  for (let i = 0; i < maxAttempts; i++) {
    const r = await client.getTxReceipt(txHash);
    if (r && r.done) return r;
    await sleep(intervalMs);
  }
  return null;
}

async function main() {
  const args = parseArgs(process.argv);
  const userKey = normalizePrivateKey(args.userKey);
  const bridge = loadBridgeAddress(args.bridge);
  const recipientHex = args.solanaRecipientBase58
    ? parseRecipientBase58(args.solanaRecipientBase58)
    : parseRecipientHex64(args.solanaRecipientHex);

  if (!Number.isFinite(args.amountSeth) || args.amountSeth < 1) {
    throw new Error('--amount-seth must be >= 1');
  }
  if (!Number.isFinite(args.minSusdcRaw) || args.minSusdcRaw < 0) {
    throw new Error('--min-susdc-raw must be >= 0');
  }

  const client = new SethClient(args.host, args.port);
  const userAddress = deriveAddressFromPrivateKey(client, userKey);
  const before = await client.getTotalWithdrawRequests(userAddress, bridge);

  console.log(`[outbound-js] Seth: ${args.host}:${args.port}`);
  console.log(`[outbound-js] Bridge: 0x${bridge}`);
  console.log(`[outbound-js] User: 0x${userAddress}`);
  console.log(`[outbound-js] totalWithdrawRequests(before): ${before}`);

  const input =
    functionSelector('requestWithdrawToSolanaFromSETH(bytes32,uint256)') +
    recipientHex +
    uint256Hex(args.minSusdcRaw);

  const tx = await client.sendTransaction(userKey, {
    to: bridge,
    step: 8,
    input,
    amount: String(args.amountSeth),
    gas_limit: args.gasLimitCall,
    gas_price: 1,
  });

  if (!tx.success || !tx.txHash) {
    throw new Error(`send failed: ${tx.error || 'unknown error'}`);
  }

  const receipt = await waitReceipt(client, tx.txHash);
  if (!receipt) throw new Error('receipt timeout');
  if (receipt.status === 5) throw new Error(`status=5 failed: ${JSON.stringify(receipt.raw)}`);

  console.log(`[outbound-js] tx: ${tx.txHash}`);
  console.log(`[outbound-js] receipt status: ${receipt.status}`);

  const after = await client.getTotalWithdrawRequests(userAddress, bridge);
  console.log(`[outbound-js] totalWithdrawRequests(after): ${after}`);
  if (!(after > before)) {
    throw new Error('request counter did not increase');
  }

  const requestId = after;
  const req = await client.getWithdrawRequest(userAddress, bridge, requestId);
  const requestKey = await client.getWithdrawRequestKey(userAddress, bridge, requestId);

  console.log(`[outbound-js] request_id: ${requestId}`);
  console.log(`[outbound-js] request: ${JSON.stringify(req)}`);
  console.log(`[outbound-js] lock/withdraw request key: ${requestKey}`);
  console.log('[outbound-js] next: ensure relayer is running to complete Solana unlock');
}

main().catch((e) => {
  console.error(`[outbound-js] failed: ${e.message}`);
  process.exit(1);
});

